form conditions fields

escopecz opened the pull request:
Documentation PR: https://github.com/mautic/mautic-documentation/pull/208

~⚠️ This PR is blocked by #9761 which needs to be merged before this PR.~

Clone of https://github.com/mautic/mautic/pull/8660 for M3

Closes #8595 

A new feature was requested for combination of Country and conditional State field. The goal is to show only the states for the selected country.

### Steps to test
- Follow https://github.com/mautic/mautic/pull/8660

### To test the Company-state field feature:
1. Create a new form
2. Create a country field
3. Create a new conditional select field for the country field. Make it visible for any country value and map it to the State contact field. The choices should be used from the contact field as well. Do not create custom ones.
4. Save the form.
5. Go to the form preview.


<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/8794"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

kuzmany committed:
Init commit

kuzmany committed:
Add conditions column to form_fields

kuzmany committed:
Adda ajax lookup for conditional options

kuzmany committed:
Adda ajax choosen for conditional field

kuzmany committed:
Added info about conditions to panel

kuzmany committed:
Start working on parent/child structure of conditional fields

kuzmany committed:
Another work for conditional fields

kuzmany committed:
Another part work on conditional field backend

kuzmany committed:
Rework backend conditinal fields with parent as string

kuzmany committed:
Added showOn

kuzmany committed:
Redesign backend for conditional fields

kuzmany committed:
Minor

kuzmany committed:
Add showForConditionalField to skip field parent not want it

kuzmany committed:
Add support for multi select values

kuzmany committed:
Add field matching and validation for conditional fields

kuzmany committed:
Fix ordering for nonparent/parent

kuzmany committed:
Add PropertiesAccessor to access to properties

kuzmany committed:
Delete children fields after parent remove

kuzmany committed:
Add field save/delete events for developers

kuzmany committed:
Rework fields styling

kuzmany committed:
Minor fixes

kuzmany committed:
Just one level conditioanl at the moment

kuzmany committed:
Add support for Any selected value

kuzmany committed:
Fix syncList from contact field

kuzmany committed:
Redesign link to create new sub field

kuzmany committed:
Minor CSS changes

kuzmany committed:
Fix reordering and improve form details  field list

kuzmany committed:
Minor to style/css

kuzmany committed:
Delete form-field.html.php

kuzmany committed:
Add toggle to show/hidden sub field select

kuzmany committed:
Add include/exclude

kuzmany committed:
Add support for checkbox and radio groups

kuzmany committed:
Fix radio/checkbox groups in backend

kuzmany committed:
Fix country selector

kuzmany committed:
Replace toggle to show for child  selector

escopecz committed:
Migration that adds form_fields.parent_id field

escopecz committed:
Fixing PHP Notice - Undefined index: empty_value

escopecz committed:
Shorter syntax used

escopecz committed:
Preparing the subscriber for M3 compatibility

Also removing unnecessary comments, applyging defensive programming technique

escopecz committed:
Removing unecessary abstract class and using the interface directly

Also adding final as it's a value object and doesn't need to be mocked in tests. Adding missing getter so all the properties from the constructor could be used, removing unecessary comments

escopecz committed:
Define class properties

escopecz committed:
The data-mautic-form-show-on param wasn't set if the $fields weren't passed to the field_helper.

escopecz committed:
Oops, the radio field already was passing the fields

escopecz committed:
Cover all posibilities of Field::showForConditionalField() with tests

escopecz committed:
Refactoring to return early

escopecz committed:
CS fixes

escopecz committed:
FormFieldEventTest added

escopecz committed:
Namespace fix for a test

escopecz committed:
FormConditionalSubscriber tested

escopecz committed:
FormFieldConditionType tested

escopecz committed:
Removing unused dependency

escopecz committed:
PropertiesAccessor test

escopecz committed:
Changes in FormModel tested

escopecz committed:
Removing the failing test for now. Discussing with the author how to cover the code

escopecz committed:
Merge branch 'form-conditions-fields' into m3-form-conditions-fields

escopecz committed:
PhpUnit tests refactored for namespaces

escopecz committed:
Automatic CS Fixer changes

escopecz committed:
signature compatibility

escopecz committed:
Namespace fix

escopecz committed:
Use PHP 7.1 features for the new classes and methods

escopecz committed:
Switching choice lables and values for Symfony 3

escopecz committed:
Replacing old exception with the new one

escopecz committed:
Switching labels with values in a select box (Symfony 2 -> 3)

alanhartless committed:
Only prepare the forms once the DOM is fully loaded so that Javascript can do it's thing in form previews, embedded on landing pages and embedded into 3rd party

alanhartless committed:
Added missing semi-colon

escopecz committed:
Wrong list of field types was shown during adding conditional fields

right after a choice field was created or edited.

escopecz committed:
Fixing PHP Notice - Undefined index: type

The if was checking whether the formfield was not null. But the form field was an empty array

escopecz committed:
Implementing special behaviour for opt group conditional select fields to make the country-state combo simple

escopecz committed:
Prod assets rebuilt

escopecz committed:
Added missing variable as it's used but not defined. Caused by cherry-picking and rebasing back and forth.

npracht commented:
Link to doc : https://docs.google.com/document/d/1JfLmimbuYPXRwxBrUcusM1-ITQLP-TyCMU1QGvG27Ss/edit?ts=5ec7d494#heading=h.vc029lot0qwv

kuzmany committed:
Stop display child If parent is hidden type

kuzmany commented:
I found issue for M2 version If parent field is hidden type (Show If value exist is disabled)

https://github.com/mautic/mautic/pull/8660/commits/e40aaf33a217dbe0b97275720d9d88bb4a45c6ad

escopecz commented:
@kuzmany thinking about this, is there even any use case why we should make hidden conditional fields? The hidden field will be in the HTML for the whole time anyway. No matter whether visible or hidden.

kuzmany commented:
@escopecz 
we have two states of hidden

1. If showForContact = false https://github.com/mautic/mautic/blob/staging/app/bundles/FormBundle/Views/Builder/form.html.php#L67
Then input is not render to form
2. If there is some combinations getShowWhenValueExists, getLeadField, getIsAutoFill ...
https://github.com/mautic/mautic/blob/staging/app/bundles/FormBundle/Views/Builder/form.html.php#L77, then is input covert to hidden type

For the second one , we don't wanna generate children field.

kuzmany commented on app/bundles/FormBundle/Views/Field/freehtml.html.php:
```diff
@@ -12,7 +12,7 @@
 include __DIR__.'/field_helper.php';
 
 if ($inBuilder) {
-    $htmlContent = $view['content']->showScriptTags($properties['text']);
```
This was removed from my commits or from yours? 
This should prevent run script code from builder
We could revert it

![image](https://user-images.githubusercontent.com/462477/84110954-2a194600-aa26-11ea-9091-7e700476611b.png)

escopecz replied:
Good catch. I will push a fix shortly together with other bug we found.

escopecz replied:
To answer your question, this is the commit that removed it: https://github.com/mautic/mautic/commit/1293c0f470b769d4c6a9b8f1a9cc364f96688de6#diff-0ffe1ecc742a8abb035d9e423721e222R15

kuzmany replied:
Yes, I found it. But didn't have time check it why I remove it. Maybe accidentaly

escopecz replied:
Fixed

escopecz committed:
ParentId can be string. Convert both to int before comparison

escopecz committed:
Returning back JS tag removal so it would not trigger in the builder.

https://github.com/mautic/mautic/pull/8794#pullrequestreview-426807273

escopecz committed:
Adding back empty options + tests

escopecz committed:
Hide conditional fields after submit

escopecz commented:
This PR is blocked by our [internal PR](https://github.com/mautic-inc/mautic-cloud/pull/838) that we cannot push to the community because it is blocked by https://github.com/mautic/mautic/pull/8828

kuzmany committed:
Add support for checkboxes

kuzmany committed:
Added support for radiogrp

kuzmany commented:
We've added support for 

- checkbox groups https://github.com/mautic/mautic/pull/8660/commits/028ce4f08e477f374e648b291cf33172c4524066
- radio groups https://github.com/mautic/mautic/pull/8660/commits/79fea12238bfc3a0a7a74915dc2a435f0846f716

Still consider add fix
- https://github.com/mautic/mautic/pull/8660/commits/e40aaf33a217dbe0b97275720d9d88bb4a45c6ad

reported here https://github.com/mautic/mautic/pull/8794#issuecomment-635214289

escopecz commented:
@kuzmany Thanks! I cherry-picked those commit and added to this PR.

fedys committed:
CS fixes, Removed unnecessary variable

RCheesley commented:
Close & open for CI

RCheesley commented:
@escopecz could you take a look at Travis errors please?

escopecz commented:
@RCheesley I added a note why it's failing into a comment. But the history of this PR is too long and it was hidden. I added it as a warning to the PR description. This PR is dependent on 2 other open PRs.

RCheesley commented:
Here's hoping we can get them tested and documented so that they can be merged in 3.1! Most are features so that's the earliest chance we have to merge them.

It would be great if the Acquia team could please take a look at their PR's at some point as many are missing documentation, have conflicts, need to be rebased etc.

RCheesley commented:
Bumping to 3.2 due to PR that blocks this not having responses to feedback. Here's hoping we can get these PR's merged in November!

RCheesley commented:
@cla-bot check

dadarya0 committed:
Remove first child selection

npracht committed:
Merge branch 'features' into m3-form-conditions-fields

kuzmany committed:
#8794 - fix showForConditionalField for 0 value in parent

npracht commented:
@kuzmany can you review it ? It would be good to merge it in 3.3.0 :)

kuzmany committed:
Add clean subscriber to values from condition

RCheesley commented:
@mautic/acquia-po @escopecz @alanhartless the internal PR should now be unblocked - would be helpful if it could be pushed over to the community so that we can get this feature released!

kuzmany committed:
Fix FormFieldConditionTypeTest

kuzmany committed:
Fix SubmissionFunctionalTest

kuzmany committed:
Fix FormFieldConditionTypeTest

kuzmany commented:
Related fixes to this PR https://github.com/escopecz/mautic/pull/17

fedys committed:
Merge branch '3.3' into m3-form-conditions-fields-MERGE

# Conflicts:
#	app/bundles/FormBundle/Controller/FieldController.php
#	app/bundles/FormBundle/Entity/Field.php
#	app/bundles/FormBundle/Views/Builder/fieldwrapper.html.php
#	app/bundles/FormBundle/Views/Builder/form.html.php

fedys committed:
Fixed SubmissionFunctionalTest

fedys committed:
CS fix

kuzmany committed:
Merge remote-tracking branch 'escopecz/m3-form-conditions-fields' into m3-form-conditions-fields-fix-types

kuzmany committed:
Fix SubmissionFunctionalTest.php

npracht committed:
Merge branch 'features' into m3-form-conditions-fields

escopecz committed:
Merge pull request #17 from kuzmany/m3-form-conditions-fields-fix-types

#8794 - fix showForConditionalField for 0 value in parent

kuzmany commented on app/bundles/CoreBundle/Assets/js/4.builder.js:
```diff
@@ -219,7 +219,7 @@ Mautic.formatCode = function() {
  */
 Mautic.openMediaManager = function() {
     Mautic.openServerBrowser(
-        mauticBasePath + (mauticEnv === 'dev' ? '/index_dev.php' : '') + '/elfinder',
```
I see here some changes what not related to this PR as like this

fedys replied:
@kuzmany this change was made by you. Shall I revert it?

kuzmany replied:
Yes

kuzmany reviewed (changes requested):
media/css/app.css 
media/js/app.js
mautic-form.js 

should not be part of this PR

RCheesley commented:
@fedys could you take a look at the feedback from @kuzmany above? Thanks!

fedys committed:
Reverted assets files:
media/css/app.css
media/js/app.js
mautic-form.js

fedys committed:
Reverted 4.builder.js

kuzmany commented on app/bundles/LeadBundle/EventListener/ReportSubscriber.php:
```diff
@@ -247,6 +247,7 @@ public function onReportGenerate(ReportGeneratorEvent $event)
                 } else {
                     $event->applyDateFilters($qb, 'date_added', 'l');
                 }
+                $event->addCompanyLeftJoin($qb);
```
This should not part of this PR

kuzmany commented on plugins/MauticFocusBundle/EventListener/CampaignSubscriber.php:
```diff
@@ -79,7 +80,7 @@ public function onCampaignTriggerAction(CampaignExecutionEvent $event)
         }
         $values                 = [];
-        $values['focus_item'][] = ['id' => $focusId, 'js' => $this->router->generate('mautic_focus_generate', ['id' => $focusId], true)];
+        $values['focus_item'][] = ['id' => $focusId, 'js' => $this->router->generate('mautic_focus_generate', ['id' => $focusId], UrlGeneratorInterface::ABSOLUTE_URL)];
```
Shouln't part of this PR

kuzmany commented on .github/workflows/mautic-asset-upload/upload.php:
```diff
@@ -76,7 +76,7 @@
     'file'            => $response['file']['name'],
     'category'        => $assetCategoryId,
-    'isPublished'     => false,
+    'isPublished'     => true,
```
Shouln't part of this PR

kuzmany commented on app/bundles/ConfigBundle/Controller/ConfigController.php:
```diff
@@ -26,31 +26,6 @@
 
 class ConfigController extends FormController
 {
-    /**
```
Shouln't part of this PR

kuzmany reviewed (changes requested):
I found more files not related to this PR, I commented it.

fedys committed:
Reverted files that are not a part of the PR

kuzmany reviewed (dismissed):
We use this on production long time. It works as describe 👍

dennisameling committed:
Merge branch 'features' into m3-form-conditions-fields

RCheesley committed:
Merge branch 'features' into m3-form-conditions-fields

RCheesley committed:
Merge branch 'features' into m3-form-conditions-fields

RCheesley commented:
@fedys @escopecz there are still files in this PR that should not be, according to the comments from @kuzmany.

Could you please take a look and address those issues and the failing tests? Thanks!

RCheesley commented:
I have written the documentation PR based on the PR from @kuzmany and the notes on this PR: https://github.com/mautic/mautic-documentation/pull/208

RCheesley committed:
Merge branch 'features' into m3-form-conditions-fields

kuzmany committed:
Fix get choices for radio group

RCheesley commented:
@escopecz @ts-navghane this one is blocking a few other PRs - could we address the conflict so we can get this merged in 4.1? Thank you in advance!

tejasync committed:
Merge remote-tracking branch 'origin/4.x' into m3-form-conditions-fields

* origin/4.x: (89 commits)
  Update Mautic versions in issue reporting template
  Fix (#10381)
  Version bump 4.0.0
  Merge pull request from GHSA-32hw-3pvh-vcvc
  Merge pull request from GHSA-86pv-95mj-7w5f
  Merge pull request from GHSA-72hm-fx78-xwhc
  Merge pull request from GHSA-rh5w-82wh-jhr8
  Merge pull request from GHSA-x7g2-wrrp-r6h3
  Update GrapesJs dependencies to latest version (#10336)
  [BUGFIX] Used FORM_PARAMS instead of BODY for parameters (#10301)
  Fix undefined placeholder php notice in edit form view (#9728)
  Update AssetsHelper to only check for current builder if Mautic is installed (#10372)
  Update config.yml
  Feature/add rte for grapesjs TPROD-215 (#10349)
  Fix for scenario where landing pages built with GrapesJS include unnecessary JS/CSS (#10360)
  Fix Transport Container Injection (#10352)
  Auto-update GrapesJS generated JS dist files (#10361)
  docs: add kuzmany as a contributor for test (#10366)
  docs: add fedys as a contributor for userTesting (#10365)
  docs: add julienWebmecanik as a contributor for code (#10364)
  ...

RCheesley committed:
Merge branch '4.x' into m3-form-conditions-fields

RCheesley committed:
Merge branch '4.x' into m3-form-conditions-fields

RCheesley commented:
Blocking PR is now merged so hopefully the tests will start to pass (we may need to pull the 4.0 branch up to the 4.x branch first) and we can start to test this PR, which is blocking a *lot* of other PRs at Acquia from coming over to the community.

RCheesley committed:
Merge branch '4.x' into m3-form-conditions-fields

RCheesley commented:
@escopecz the PR this was dependent on is now merged, and should be within this branch. Hopefully we can get cracking with testing this PR once the tests are passing.

kuzmany commented on .github/workflows/mautic-asset-upload/upload.php:
```diff
@@ -76,7 +76,7 @@
     'file'            => $response['file']['name'],
     'category'        => $assetCategoryId,
-    'isPublished'     => true,
+    'isPublished'     => false,
```
Should not be this part of this PR  I think

escopecz replied:
@fedys I'm also confused what is it doing here. I can see it's coming from https://github.com/mautic/mautic/commit/d983a889681096a25f2674f6ed63689ddf337add. Do you remember what is the story here?

fedys replied:
@escopecz I think I was just fixing some PR that included unrelated changes.

kuzmany replied:
There are more changes not related to this PR for example:

https://github.com/mautic/mautic/pull/8794/files#diff-e192a95a44e25a558822230a82b49a099f01a1f2a6ffecf274b06b1579e7a94cR36
https://github.com/mautic/mautic/pull/8794/files#diff-fb63f85c5c793dbfedafce713ba68125484bd35d7cab18c3e0493a359a26cb8bL222

escopecz committed:
Merge branch '4.x' into m3-form-conditions-fields

escopecz committed:
The at() matcher has been deprecated.

escopecz committed:
Reverting changes from a wild commit

https://github.com/mautic/mautic/pull/8794/commits/d983a889681096a25f2674f6ed63689ddf337add

These changes shouldn't be part of https://github.com/mautic/mautic/pull/8794

escopecz commented:
@kuzmany I reverted changes that shouldn't be in this PR from https://github.com/mautic/mautic/pull/8794/commits/d983a889681096a25f2674f6ed63689ddf337add by https://github.com/mautic/mautic/pull/8794/commits/00a84d79a83b85b3a21dcbfef19bd72d81436ecf

kuzmany commented:
@escopecz thanks, I will check it. Don't forget we have another fix in your repo https://github.com/escopecz/mautic/pull/18

escopecz committed:
Merge pull request #18 from kuzmany/m3-form-conditions-fields-fix-choices-radiogroup

Fix get choices for radio group

escopecz committed:
Trying to run the tests failing in CI in a separate process if it helps

escopecz committed:
Merge branch 'm3-form-conditions-fields' of github.com:escopecz/mautic into m3-form-conditions-fields

escopecz committed:
Reverting back as the previous change did not help

escopecz commented:
I'm stuck on resolving the failing tests here. Here are my observations:
1. The tests are failing only in GitHub Actions.
2. The tests  pass fine on my local if I run only the specific test as well as if I run the whole test suite.
3. It was passing yesterday on https://github.com/mautic/mautic/pull/8794/commits/05171eea6b264af911253b31f15b1e56ef7e0f57 but it started to fail on commit that removed some code that was not related to this PR (https://github.com/mautic/mautic/pull/8794/commits/00a84d79a83b85b3a21dcbfef19bd72d81436ecf). I cannot find any link between the failing test and removed code.
4. I tried to run the failing tests in a separate process as it sometimes helps in these cases, but it did not.

@fedys @dennisameling would you be able to check if you'll get more ideas what else to try?

escopecz committed:
Merge branch '4.x' into m3-form-conditions-fields

escopecz committed:
Adding checks and debug info to problematic tests

escopecz committed:
Using dynamic campaign ID in a failing test

escopecz committed:
Adding more details to the failing test

escopecz committed:
Using dynamic IDs for contacts in a fragile test

escopecz committed:
Make email unique for the test as if John Doe was ever created by another test before this test will fail.

escopecz committed:
Use dynamic IDs in a fragile test

escopecz committed:
One more place where to use dynamic ID

escopecz committed:
Replacing forgotten hard-coded ID

escopecz committed:
Select the right audit log record

escopecz commented:
The failing tests need fixes from https://github.com/mautic/mautic/pull/10470, so waiting for that to be merged first.

escopecz committed:
PHPSTAN fixes

escopecz committed:
Merge branch 'm3-form-conditions-fields' of github.com:escopecz/mautic into m3-form-conditions-fields

escopecz committed:
PHPSTAN fixes

escopecz committed:
CS fixes

escopecz committed:
Improving fragile test

escopecz committed:
More debug messages added to the test

escopecz committed:
Fixing undefined array prop

kuzmany commented on app/bundles/CoreBundle/Assets/js/4.builder.js:
```diff
@@ -219,7 +219,7 @@ Mautic.formatCode = function() {
 Mautic.openMediaManager = function() {
     Mautic.openServerBrowser(
-        mauticBasePath + (typeof mauticEnv !== 'undefined' &&  mauticEnv === 'dev' ? '/index_dev.php' : '') + '/elfinder',
+        mauticBasePath + (mauticEnv === 'dev' ? '/index_dev.php' : '') + '/elfinder',
```
This return error with image insert by legacy builder

![image](https://user-images.githubusercontent.com/462477/141692783-89eb1d73-57fb-4333-b86b-8eb302e04a3c.png)

escopecz replied:
Good catch @kuzmany! Fixed in https://github.com/mautic/mautic/pull/8794/commits/0206837cd4bac7f1ded577da369d24ac71978b09

kuzmany reviewed (dismissed):
I did basic test and works. 👍 
Code looks 👍 
Covered by unit tests 👍 
Let's go 👍 

I see just issue in app/bundles/CoreBundle/Assets/js/4.builder.js

escopecz committed:
Reverting bad merge change

dennisameling commented:
> The failing tests need fixes from #10470, so waiting for that to be merged first.

@escopecz FYI #10470 has just been merged, so that should unblock this PR 👍🏼

escopecz commented:
Thanks @dennisameling! I cherry-picked the commits I needed earlier today. So this PR is not failing anymore.

RCheesley committed:
Merge branch '4.x' into m3-form-conditions-fields

kuzmany reviewed (dismissed):
Very very cool PR 
Tested, unit tests included 
Then, let's go 👍

RCheesley commented:
@escopecz looks like we have a conflict to address here now.

bell87 reviewed (dismissed):
Tested, works well.

RCheesley commented:
GTG as soon as the conflict is resolved! 🎉

escopecz committed:
Merge branch '4.x' into m3-form-conditions-fields

escopecz commented:
Conflict resolved

RCheesley commented:
Thanks for all the work on this PR and persistence over the last year folks!

RCheesley commented:
@all-contributors please add @bell87 for usertesting
